src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+circpad_distribution_sample(circpad_distribution_t dist)
+{
+  double p = 0;
+
+  switch (dist.type) {
+    case CIRCPAD_DIST_NONE:
+      return 0;
+    case CIRCPAD_DIST_UNIFORM:
+      p = crypto_rand_double();
+      // param2 is upper bound, param1 is lower
+      /* The subtraction is exact as long as param2 and param1 are less than
+       * 2**53. The multiplication is accurate as long as (param2 - param1)
+       * is less than 2**52. (And when they are large, the low bits aren't
+       * important.) The result covers the full range of outputs, as long as
+       * p has a resolution of 1/2**32 or greater. */
+      p *= (dist.param2 - dist.param1);
+      p += dist.param1;
+      return p;
+    case CIRCPAD_DIST_LOGISTIC:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Logistic_distribution#Quantile_function
+       * param1 is Mu, param2 is s. */
+      if (p <= 0.0) // Avoid log(0)
+        return 0;
+      return dist.param1 + dist.param2*tor_mathlog(p/(1.0-p));
+    case CIRCPAD_DIST_LOG_LOGISTIC:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Log-logistic_distribution#Quantiles
+       * param1 is Alpha, param2 is Beta */
+      return dist.param1 * pow(p/(1.0-p), 1.0/dist.param2);
+    case CIRCPAD_DIST_GEOMETRIC:
+      p = crypto_rand_double();
+      /* https://github.com/distributions-io/geometric-quantile/
+       * param1 is 'p' (success probability) */
+      return ceil(tor_mathlog(1.0-p)/tor_mathlog(1.0-dist.param1));
+    case CIRCPAD_DIST_WEIBULL:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Weibull_distribution \
+       *    #Cumulative_distribution_function
+       * param1 is k, param2 is Lambda */
+      return dist.param2*pow(-tor_mathlog(1.0-p), 1.0/dist.param1);
+    case CIRCPAD_DIST_PARETO:
+      p = 1.0-crypto_rand_double(); // Pareto quantile needs (0,1]
+
+      /* https://en.wikipedia.org/wiki/Generalized_Pareto_distribution \
+       *    #Generating_generalized_Pareto_random_variables
+       * param1 is Sigma, param2 is Xi
+       * Since it's piecewise, we must define it for 0 (or close to 0) */
+      if (fabs(dist.param2) <= 1e-22)
+        return -dist.param1*tor_mathlog(p);
+      else
+        return dist.param1*(pow(p, -dist.param2) - 1.0)/dist.param2;
+  }
+  return 0;
+}
+
+/**
+ * Find the index of the first bin whose upper bound is
+ * greater than the target, and that has tokens remaining.
+ */
+static int
+circpad_machine_first_higher_index(circpad_machineinfo_t *mi,
+                                   circpad_delay_t target_bin_us)
+{
+  int i = circpad_histogram_usec_to_bin(mi, target_bin_us);
+
+  /* Don't remove from the infinity bin */
+  for (; i < mi->histogram_len-1; i++) {
+    if (mi->histogram[i] &&
+        circpad_histogram_bin_to_usec(mi, i+1) > target_bin_us) {
+      return i;
+    }
+  }
+
+  return mi->histogram_len;
+}
+
+/**
+ * Find the index of the first bin whose lower bound is
```

Comment:
We need to fix the documentation (and maybe naming) of this function and of `circpad_machine_remove_higher_token` since they dont return the first bin that is `lower`, they return the first bin that is `lower-or-equal` to the target.

Reply:
Addressed in 75b7da0ced.

Reply:
Done in 0a4e04a84f0b18343bd38732c78951c295c73a1d.